src/python/zquantum/core/interfaces/functions.py

```diff
@@ -43,10 +45,49 @@ def gradient(self, params: np.ndarray) -> np.ndarray:
         pass
 
 
+def has_store_artifact_param(function) -> bool:
+    """Determine if given callable is capable of storing artifacts.
+
+    :param function: a callable to be checked.
+    :return: True, if `function` has store_artifact parameter and False otherwise.
+    """
+    try:
+        return "store_artifact" in signature(function, follow_wrapped=True).parameters
+    except ValueError:
+        # Rationale: the only callables that are of interest to us that aren't supported by
+        # signature that I am aware of are numpy ufunc's. Obviously, they don't have
+        # store_artifact parameter.
+        return False
+
+
 class FunctionWithGradient(NamedTuple):
     """A callable with gradient."""
     function: Callable[[np.ndarray], np.ndarray]
     gradient: Callable[[np.ndarray], np.ndarray]
 
     def __call__(self, params: np.ndarray):
         return self.function(params)
+
+
+class FunctionWithGradientStoringArtifacts(NamedTuple):
+    """A callable with gradient that also stores artifacts."""
+    function: CallableStoringArtifacts
+    gradient: Callable[[np.ndarray], np.ndarray]
+
+    def __call__(self, params: np.ndarray, store_artifact: StoreArtifact):
+        return self.function(params, store_artifact)
+
+
+def function_with_gradient(
```

Comment:
Question about the naming convention you're using here:
How come we use CamelCase for `FunctionWithGradient`, yet snake_case for `function_with_gradient`?

Reply:
Oh, that's because `FunctionWithGradient` is a class and `function_with_gradient` is a function (I guess one could call it a factory for `FunctionWithGradient` and `FunctionWithGradientStoringArtifacts`. That's the same convention as in `recorder.py` where actually the `recorder` is a factory for several of history-recorder classes.